mesonbuild/cmake/interpreter.py

```diff
@@ -1183,6 +1183,11 @@ def process_target(tgt: ConverterTarget) -> None:
                 else:
                     dep_var = None
 
+            if dep_var:
+                meson = id_node('meson')
+                node = method(meson, 'override_dependency', args=[tgt.cmake_name, id_node(dep_var)])
+                node_list += [node]
+
```

Comment:
-1 to automatically adding dependency overrides, we can't assume that every target maps to a publicly exported pkg-config equivalent. We don't do this for meson subprojects either!

If users want to mark a dependency as an override in meson subprojects, then someone needs to handcraft that relationship -- whether that means using meson.override_dependency() or using a `[provide]` mapping in the wrap file, is just semantics and doesn't really matter.

This patch is would make cmake violate that semantic.

Reply:
Hm, you're right that's done for every lib and not only public libraries. The problem is a cmake .wrap file cannot use the `[provide] dep = dep_var` syntax because the name of the variable is generated and should not guarantee that name IMHO.

What about passing the list from `[provide] dependency_names = foo, bar` to CMakeInterpreter in `_do_subproject_cmake()`, and change the condition above to `if dep_var and tgt.cmake_name in self.provided_deps:`? That way having a dependency in the wrap `[provide]` makes the generated AST override them.

Reply:
Even that won't work, because it's pretty common for the literal library name to be different from the pkg-config name.

Why isn't the name of the variable guaranteed?

We'd need to be able to map a dependency name to a target name in some manner in order to guarantee semantics, and that's what users of `cmake.subproject()` do with `depname = subp.dependency('target_name')`.

Reply:
> Even that won't work, because it's pretty common for the literal library name to be different from the pkg-config name.

It does not have to match, that's why we have e.g. `dependency('foo', 'FOO')` to check multiple names. The cmake name is often different to pkgconfig name.

> Why isn't the name of the variable guaranteed?

We could, but I really dislike hardcoding variable names in wrap file, I've always wanted to deprecate that.

> that's what users of cmake.subproject() do with depname = subp.dependency('target_name').

'target_name' there can be any cmake target, that does not mean it's a public dependency or a pkgconfig name neither.

How does cmake define the public name of a dependency? I thought that was simply the target name.

Reply:
Cmake doesn't have a public name for a dependency, you are expected to read CMakeLists.txt and use different values depending on whether you are using system deps via pkg-config (discouraged due to NIH), system deps via find_package, or vendored deps via building a subproject.

Due to cmake being cmake, you even need to know a variable name for find_package, and then a different name for the global variable that find_package evaluates and exports. Meson erases this difference if it can, by checking how many variables a find_package exports and in non-ambiguous cases of only one exported target, using that.

Often times the target name has the word "static" in it.

Cmake is basically designed to be confusing here.

Reply:
> that's why we have e.g. `dependency('foo', 'FOO')` to check multiple names. The cmake name is often different to pkgconfig name.

Yeah, and that's a bug. Also we run pkg-config once per name. Even worse, we build a dummy cmake project once per name. Adding more names isn't the solution here.

Reply:
Hmm, this all sucks... So what you suggest is having `[provide] the_public_name = <target_name>_dep` in wrap and we guarantee that all our declare_dependency() have that variable name?

Reply:
Ok, removed override_dependency PR and added notes in the doc that the provide section must use the syntax above.